Bug Report: Invalid use of group function for sharded queries

### Overview of the Issue

Using Vitess 14 I'm getting `Invalid use of group function` for a query against a sharded db (works on unsharded): `SELECT sse.id, sse.portalId, sse.sequenceId, sse.enrollmentId, sse.sequenceStepId, sse.state, sse.stepOrder, sse.actionType, sse.delayMillis, sse.scheduledAtUpdatedAt, sse.scheduledAt, sse.timeOfDayMillis, sse.createdAt, sse.updatedAt, sse.trackerKey FROM sequence_step_enrollments AS sse JOIN (SELECT sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) AS stepOrder FROM sequence_step_enrollments AS sub WHERE sub.portalId = :portalId AND sub.enrollmentId IN (<enrollmentIds>) GROUP BY sub.portalId, sub.enrollmentId ORDER BY NULL) AS maxStep ON maxStep.portalId = sse.portalId AND maxStep.enrollmentId = sse.enrollmentId AND maxStep.stepOrder = sse.stepOrder WHERE sse.portalId = :portalId AND sse.enrollmentId IN (<enrollmentIds>)`

the query it seems to actually be executing is `select /*+ MAX_EXECUTION_TIME(30000) */ 1 from (select sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) as stepOrder from sequence_step_enrollments as sub where sub.portalId = :v1 and sub.enrollmentId in (:v2, :v3) and sub.portalId = :sse_portalId and sub.enrollmentId = :sse_enrollmentId and MAX(sub.stepOrder) = :sse_stepOrder group by sub.portalId, sub.enrollmentId order by null) as maxStep`

It works with the v3 planner but not gen4.

I tried rewriting the `MAX(sub.stepOrder)` as (`SELECT MAX(stepOrder) FROM ...)` but then I get `unknown type *physical.SubQueryOp pushJoinPredicates`

Slack convo: https://vitess.slack.com/archives/C0PQY0PTK/p1663785636835939

### Reproduction Steps

vschema:
```
{
  "sharded": true,
  "vindexes": {
    "hash": {
      "type": "hash"
    },
    "null": {
      "type": "null"
    }
  },
  "tables": {
    "sequence_step_enrollments": {
      "columnVindexes": [
        {
          "column": "portalId",
          "name": "hash"
        }
      ],
      "autoIncrement": {
        "column": "id",
        "sequence": "SequencesV2__sequence_step_enrollments__id__seq"
      }
    }
  }
}
```

schema:
```
mysql> show create table sequence_step_enrollments\G
*************************** 1. row ***************************
       Table: sequence_step_enrollments
Create Table: CREATE TABLE `sequence_step_enrollments` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `portalId` int(10) unsigned NOT NULL,
  `sequenceId` bigint(20) unsigned NOT NULL,
  `enrollmentId` bigint(20) unsigned NOT NULL,
  `sequenceStepId` bigint(20) unsigned DEFAULT NULL,
  `state` tinyint(3) unsigned NOT NULL,
  `stepOrder` tinyint(3) unsigned NOT NULL,
  `actionType` tinyint(3) unsigned NOT NULL,
  `delayMillis` bigint(20) NOT NULL,
  `scheduledAtUpdatedAt` bigint(20) unsigned NOT NULL,
  `scheduledAt` bigint(20) unsigned NOT NULL,
  `createdAt` bigint(20) unsigned NOT NULL,
  `updatedAt` bigint(20) unsigned NOT NULL,
  `trackerKey` varchar(42) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `timeOfDayMillis` bigint(20) unsigned NOT NULL DEFAULT '0',
  PRIMARY KEY (`id`),
  UNIQUE KEY `idx_portalId_enrollmentId_stepOrder` (`portalId`,`enrollmentId`,`stepOrder`),
  KEY `idx_portalId_trackerKey` (`portalId`,`trackerKey`),
  KEY `idx_portalId_enrollmentId_state_actionType_scheduledAt` (`portalId`,`enrollmentId`,`state`,`actionType`,`scheduledAt`),
  KEY `idx_portalId_enrollmentId` (`portalId`,`enrollmentId`)
) ENGINE=InnoDB AUTO_INCREMENT=1628433300 DEFAULT CHARSET=utf8mb4 ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=8
```

Run query
```
SELECT sse.id, sse.portalId, sse.sequenceId, sse.enrollmentId, sse.sequenceStepId, sse.state, sse.stepOrder, sse.actionType, sse.delayMillis, sse.scheduledAtUpdatedAt, sse.scheduledAt, sse.timeOfDayMillis, sse.createdAt, sse.updatedAt, sse.trackerKey FROM sequence_step_enrollments AS sse JOIN (SELECT sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) AS stepOrder FROM sequence_step_enrollments AS sub WHERE sub.portalId = 123 AND sub.enrollmentId IN (2) GROUP BY sub.portalId, sub.enrollmentId ORDER BY NULL) AS maxStep ON maxStep.portalId = sse.portalId AND maxStep.enrollmentId = sse.enrollmentId AND maxStep.stepOrder = sse.stepOrder WHERE sse.portalId = :portalId AND sse.enrollmentId IN (2)
```

### Binary Version

```sh
Version: 14.0.2-SNAPSHOT (Jenkins build 1)
```


### Operating System and Environment details

```sh
Can repro in both linux and macos
```


### Log Fragments

```sh
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: 
java.sql.SQLSyntaxErrorException: message: "target: test_db_MgP1Tv.0.primary: vttablet: rpc error: code = InvalidArgument desc = Invalid use of group function (errno 1111) (sqlstate HY000) (CallerID: unsecure_grpc_client): Sql: \"select /*+ MAX_EXECUTION_TIME(30000) */ 1 from (select sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) as stepOrder from sequence_step_enrollments as sub where sub.portalId = :v1 and sub.enrollmentId in (:v2, :v3) and sub.state in (:v4, :v5, :v6, :v7, :v8, :v9, :v10, :v11, :v12) and sub.portalId = :sse_portalId and sub.enrollmentId = :sse_enrollmentId and MAX(sub.stepOrder) = :sse_stepOrder group by sub.portalId, sub.enrollmentId order by null) as maxStep /* user:unsecure_grpc_client */\", BindVars: {#maxLimit: \"type:INT64 value:\\\"10001\\\"\"sse_enrollmentId: \"type:UINT64 value:\\\"6670104721698623488\\\"\"sse_portalId: \"type:UINT32 value:\\\"1553004775\\\"\"sse_stepOrder: \"type:UINT8 value:\\\"1\\\"\"v1: \"type:INT64 value:\\\"1553004775\\\"\"v10: \"type:INT64 value:\\\"7\\\"\"v11: \"type:INT64 value:\\\"8\\\"\"v12: \"type:INT64 value:\\\"9\\\"\"v13: \"type:INT64 value:\\\"1553004775\\\"\"v14: \"type:INT64 value:\\\"6670104721698623488\\\"\"v15: \"type:INT64 value:\\\"9139429175885435904\\\"\"v16: \"type:INT64 value:\\\"1\\\"\"v17: \"type:INT64 value:\\\"2\\\"\"v18: \"type:INT64 value:\\\"3\\\"\"v19: \"type:INT64 value:\\\"4\\\"\"v2: \"type:INT64 value:\\\"6670104721698623488\\\"\"v20: \"type:INT64 value:\\\"5\\\"\"v21: \"type:INT64 value:\\\"6\\\"\"v22: \"type:INT64 value:\\\"7\\\"\"v23: \"type:INT64 value:\\\"8\\\"\"v24: \"type:INT64 value:\\\"9\\\"\"v3: \"type:INT64 value:\\\"9139429175885435904\\\"\"v4: \"type:INT64 value:\\\"1\\\"\"v5: \"type:INT64 value:\\\"2\\\"\"v6: \"type:INT64 value:\\\"3\\\"\"v7: \"type:INT64 value:\\\"4\\\"\"v8: \"type:INT64 value:\\\"5\\\"\"v9: \"type:INT64 value:\\\"6\\\"\"}"
code: INVALID_ARGUMENT
 [statement:"SELECT sse.id, sse.portalId, sse.sequenceId, sse.enrollmentId, sse.sequenceStepId, sse.state, sse.stepOrder, sse.actionType, sse.delayMillis, sse.scheduledAtUpdatedAt, sse.scheduledAt, sse.timeOfDayMillis, sse.createdAt, sse.updatedAt, sse.trackerKey FROM sequence_step_enrollments AS sse JOIN (SELECT sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) AS stepOrder FROM sequence_step_enrollments AS sub WHERE sub.portalId = :portalId AND sub.enrollmentId IN (<enrollmentIds>) AND sub.state IN (<stepEnrollmentStates>)GROUP BY sub.portalId, sub.enrollmentId ORDER BY NULL) AS maxStep ON maxStep.portalId = sse.portalId AND maxStep.enrollmentId = sse.enrollmentId AND maxStep.stepOrder = sse.stepOrder WHERE sse.portalId = :portalId AND sse.enrollmentId IN (<enrollmentIds>) AND sse.state IN (<stepEnrollmentStates>) ", located:"SELECT sse.id, sse.portalId, sse.sequenceId, sse.enrollmentId, sse.sequenceStepId, sse.state, sse.stepOrder, sse.actionType, sse.delayMillis, sse.scheduledAtUpdatedAt, sse.scheduledAt, sse.timeOfDayMillis, sse.createdAt, sse.updatedAt, sse.trackerKey FROM sequence_step_enrollments AS sse JOIN (SELECT sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) AS stepOrder FROM sequence_step_enrollments AS sub WHERE sub.portalId = :portalId AND sub.enrollmentId IN ((:__enrollmentIds_0_0),(:__enrollmentIds_1_0)) AND sub.state IN (:__stepEnrollmentStates_0,:__stepEnrollmentStates_1,:__stepEnrollmentStates_2,:__stepEnrollmentStates_3,:__stepEnrollmentStates_4,:__stepEnrollmentStates_5,:__stepEnrollmentStates_6,:__stepEnrollmentStates_7,:__stepEnrollmentStates_8)GROUP BY sub.portalId, sub.enrollmentId ORDER BY NULL) AS maxStep ON maxStep.portalId = sse.portalId AND maxStep.enrollmentId = sse.enrollmentId AND maxStep.stepOrder = sse.stepOrder WHERE sse.portalId = :portalId AND sse.enrollmentId IN ((:__enrollmentIds_0_0),(:__enrollmentIds_1_0)) AND sse.state IN (:__stepEnrollmentStates_0,:__stepEnrollmentStates_1,:__stepEnrollmentStates_2,:__stepEnrollmentStates_3,:__stepEnrollmentStates_4,:__stepEnrollmentStates_5,:__stepEnrollmentStates_6,:__stepEnrollmentStates_7,:__stepEnrollmentStates_8) ", rewritten:"SELECT /*+ MAX_EXECUTION_TIME(30000) */ sse.id, sse.portalId, sse.sequenceId, sse.enrollmentId, sse.sequenceStepId, sse.state, sse.stepOrder, sse.actionType, sse.delayMillis, sse.scheduledAtUpdatedAt, sse.scheduledAt, sse.timeOfDayMillis, sse.createdAt, sse.updatedAt, sse.trackerKey FROM sequence_step_enrollments AS sse JOIN (SELECT sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) AS stepOrder FROM sequence_step_enrollments AS sub WHERE sub.portalId = ? AND sub.enrollmentId IN ((?),(?)) AND sub.state IN (?,?,?,?,?,?,?,?,?)GROUP BY sub.portalId, sub.enrollmentId ORDER BY NULL) AS maxStep ON maxStep.portalId = sse.portalId AND maxStep.enrollmentId = sse.enrollmentId AND maxStep.stepOrder = sse.stepOrder WHERE sse.portalId = ? AND sse.enrollmentId IN ((?),(?)) AND sse.state IN (?,?,?,?,?,?,?,?,?) ", arguments:{ positional:{}, named:{__stepEnrollmentStates_5:6,__stepEnrollmentStates_6:7,__stepEnrollmentStates_7:8,portalId:1553004775,__stepEnrollmentStates_8:9,__stepEnrollmentStates_1:2,__stepEnrollmentStates_2:3,__enrollmentIds_1_0:9139429175885435904,__stepEnrollmentStates_3:4,__enrollmentIds_0_0:6670104721698623488,__stepEnrollmentStates_4:5,__stepEnrollmentStates_0:1}, finder:[]}]
	at com.hubspot.sequences.data.v2.sql.daos.sequencestepenrollments.SequenceStepEnrollmentsDaoTest.itFetchesMostRecentByState(SequenceStepEnrollmentsDaoTest.java:1329)
Caused by: java.sql.SQLSyntaxErrorException: 
message: "target: test_db_MgP1Tv.0.primary: vttablet: rpc error: code = InvalidArgument desc = Invalid use of group function (errno 1111) (sqlstate HY000) (CallerID: unsecure_grpc_client): Sql: \"select /*+ MAX_EXECUTION_TIME(30000) */ 1 from (select sub.portalId, sub.enrollmentId, MAX(sub.stepOrder) as stepOrder from sequence_step_enrollments as sub where sub.portalId = :v1 and sub.enrollmentId in (:v2, :v3) and sub.state in (:v4, :v5, :v6, :v7, :v8, :v9, :v10, :v11, :v12) and sub.portalId = :sse_portalId and sub.enrollmentId = :sse_enrollmentId and MAX(sub.stepOrder) = :sse_stepOrder group by sub.portalId, sub.enrollmentId order by null) as maxStep /* user:unsecure_grpc_client */\", BindVars: {#maxLimit: \"type:INT64 value:\\\"10001\\\"\"sse_enrollmentId: \"type:UINT64 value:\\\"6670104721698623488\\\"\"sse_portalId: \"type:UINT32 value:\\\"1553004775\\\"\"sse_stepOrder: \"type:UINT8 value:\\\"1\\\"\"v1: \"type:INT64 value:\\\"1553004775\\\"\"v10: \"type:INT64 value:\\\"7\\\"\"v11: \"type:INT64 value:\\\"8\\\"\"v12: \"type:INT64 value:\\\"9\\\"\"v13: \"type:INT64 value:\\\"1553004775\\\"\"v14: \"type:INT64 value:\\\"6670104721698623488\\\"\"v15: \"type:INT64 value:\\\"9139429175885435904\\\"\"v16: \"type:INT64 value:\\\"1\\\"\"v17: \"type:INT64 value:\\\"2\\\"\"v18: \"type:INT64 value:\\\"3\\\"\"v19: \"type:INT64 value:\\\"4\\\"\"v2: \"type:INT64 value:\\\"6670104721698623488\\\"\"v20: \"type:INT64 value:\\\"5\\\"\"v21: \"type:INT64 value:\\\"6\\\"\"v22: \"type:INT64 value:\\\"7\\\"\"v23: \"type:INT64 value:\\\"8\\\"\"v24: \"type:INT64 value:\\\"9\\\"\"v3: \"type:INT64 value:\\\"9139429175885435904\\\"\"v4: \"type:INT64 value:\\\"1\\\"\"v5: \"type:INT64 value:\\\"2\\\"\"v6: \"type:INT64 value:\\\"3\\\"\"v7: \"type:INT64 value:\\\"4\\\"\"v8: \"type:INT64 value:\\\"5\\\"\"v9: \"type:INT64 value:\\\"6\\\"\"}"
code: INVALID_ARGUMENT
```
```

--- a/go/vt/vtgate/planbuilder/physical/route_planning.go
+++ b/go/vt/vtgate/planbuilder/physical/route_planning.go
@@ -635,7 +635,7 @@ func tryMerge(
 			return merger(aRoute, bRoute)
 		}
 	case engine.EqualUnique:
-		// if they are already both being sent to the same shard, we can merge
+		// If the two routes fully match, they can be merged together.
 		if bRoute.RouteOpCode == engine.EqualUnique {
 			aVdx := aRoute.SelectedVindex()
 			bVdx := bRoute.SelectedVindex()
@@ -644,9 +644,12 @@ func tryMerge(
 			if aVdx == bVdx && gen4ValuesEqual(ctx, aExpr, bExpr) {
 				return merger(aRoute, bRoute)
 			}
-			return nil, nil
 		}
+
+		// If the two routes don't match, fall through to the next case and see if we
+		// can merge via join predicates instead.
 		fallthrough
+
 	case engine.Scatter, engine.IN:
 		if len(joinPredicates) == 0 {
 			// If we are doing two Scatters, we have to make sure that the

--- a/go/vt/vtgate/planbuilder/testdata/select_cases.json
+++ b/go/vt/vtgate/planbuilder/testdata/select_cases.json
@@ -5416,5 +5416,47 @@
         "Table": "dual"
       }
     }
+  },
+  {
+    "comment": "Join across multiple tables, with conditions on different vindexes, but mergeable through join predicates",
+    "query": "SELECT user.id FROM user INNER JOIN music_extra ON user.id = music_extra.user_id INNER JOIN music ON music_extra.user_id = music.user_id WHERE user.id = 123 and music.id = 456",
+    "v3-plan": {
+      "QueryType": "SELECT",
+      "Original": "SELECT user.id FROM user INNER JOIN music_extra ON user.id = music_extra.user_id INNER JOIN music ON music_extra.user_id = music.user_id WHERE user.id = 123 and music.id = 456",
+      "Instructions": {
+        "OperatorType": "Route",
+        "Variant": "EqualUnique",
+        "Keyspace": {
+          "Name": "user",
+          "Sharded": true
+        },
+        "FieldQuery": "select `user`.id from `user` join music_extra on `user`.id = music_extra.user_id join music on music_extra.user_id = music.user_id where 1 != 1",
+        "Query": "select `user`.id from `user` join music_extra on `user`.id = music_extra.user_id join music on music_extra.user_id = music.user_id where `user`.id = 123 and music.id = 456",
+        "Table": "`user`, music_extra, music",
+        "Values": [
+          "INT64(123)"
+        ],
+        "Vindex": "user_index"
+      }
+    },
+    "gen4-plan": {
+      "QueryType": "SELECT",
+      "Original": "SELECT user.id FROM user INNER JOIN music_extra ON user.id = music_extra.user_id INNER JOIN music ON music_extra.user_id = music.user_id WHERE user.id = 123 and music.id = 456",
+      "Instructions": {
+        "OperatorType": "Route",
+        "Variant": "EqualUnique",
+        "Keyspace": {
+          "Name": "user",
+          "Sharded": true
+        },
+        "FieldQuery": "select `user`.id from `user`, music_extra, music where 1 != 1",
+        "Query": "select `user`.id from `user`, music_extra, music where music.id = 456 and `user`.id = 123 and `user`.id = music_extra.user_id and music_extra.user_id = music.user_id",
+        "Table": "`user`, music, music_extra",
+        "Values": [
+          "INT64(123)"
+        ],
+        "Vindex": "user_index"
+      }
+    }
   }
 ]

